File: backend/app/schemas/assistant.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.file_asset import FileUploadResponse

AssistantMode = Literal["general", "course_qa", "file_qa"]
AssistantRole = Literal["user", "assistant", "system"]
AssistantAnswerStyle = Literal["normal", "step_by_step", "concise", "detailed"]


class AssistantReference(BaseModel):
    source_type: Literal["course_knowledge", "file"]
    title: str | None = None
    filename: str | None = None
    excerpt: str
    score: float | None = None
    document_id: int | None = None
    file_id: int | None = None
    chunk_index: int | None = None


class AssistantUsedDocument(BaseModel):
    title: str | None = None
    filename: str | None = None
    source_type: Literal["course_knowledge", "file"]


class AssistantMessageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    session_id: int
    role: str
    content: str
    status: str = "completed"
    error_message: str | None = None
    completed_at: datetime | None = None
    references: list[AssistantReference] = Field(default_factory=list)
    attachment_file_ids: list[int] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class AssistantSessionSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    course_id: int | None = None
    title: str
    mode: str
    status: str
    last_message: str | None = None
    message_count: int = 0
    created_at: datetime
    updated_at: datetime


class AssistantSessionListResponse(BaseModel):
    items: list[AssistantSessionSummary]
    total: int
    page: int
    page_size: int


class AssistantSessionDetail(AssistantSessionSummary):
    messages: list[AssistantMessageRead] = Field(default_factory=list)


class AssistantSessionCreate(BaseModel):
    course_id: int | None = Field(default=None, gt=0)
    title: str | None = Field(default=None, min_length=1, max_length=160)
    mode: AssistantMode = "general"


class AssistantSendMessageRequest(BaseModel):
    message: str = Field(min_length=1, max_length=4000)
    course_id: int | None = Field(default=None, gt=0)
    use_course_knowledge: bool = True
    knowledge_document_ids: list[int] = Field(default_factory=list)
    attachment_file_ids: list[int] = Field(default_factory=list)
    answer_style: AssistantAnswerStyle = "normal"
    top_k: int = Field(default=5, ge=1, le=10)


class AssistantSendMessageResponse(BaseModel):
    session: AssistantSessionSummary
    user_message: AssistantMessageRead
    assistant_message: AssistantMessageRead
    answer: str
    references: list[AssistantReference] = Field(default_factory=list)
    used_documents: list[AssistantUsedDocument] = Field(default_factory=list)
    suggested_followups: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


class AssistantFileUploadResponse(FileUploadResponse):
    pass


class AssistantDeleteResponse(BaseModel):
    session_id: int
    deleted: bool = True

File: frontend/src/types/assistant.ts
export type AssistantMode = 'general' | 'course_qa' | 'file_qa'
export type AssistantRole = 'user' | 'assistant' | 'system'
export type AssistantAnswerStyle = 'normal' | 'step_by_step' | 'concise' | 'detailed'

export interface AssistantReference {
  source_type: 'course_knowledge' | 'file'
  title?: string | null
  filename?: string | null
  excerpt: string
  score?: number | null
  document_id?: number | null
  file_id?: number | null
  chunk_index?: number | null
}

export interface AssistantUsedDocument {
  title?: string | null
  filename?: string | null
  source_type: 'course_knowledge' | 'file'
}

export interface AssistantMessage {
  id: number
  session_id: number
  role: AssistantRole | string
  content: string
  status?: 'running' | 'completed' | 'failed' | 'cancelled' | string
  error_message?: string | null
  completed_at?: string | null
  references: AssistantReference[]
  attachment_file_ids: number[]
  created_at: string
  updated_at: string
}

export type AssistantStreamEvent =
  | { type: 'meta'; stream_supported: boolean; user_message: AssistantMessage; assistant_message: AssistantMessage; warnings: string[] }
  | { type: 'delta'; text: string }
  | { type: 'references'; references: AssistantReference[] }
  | { type: 'warning'; message: string }
  | { type: 'done'; message: AssistantMessage; references: AssistantReference[] }
  | { type: 'error'; message: string; retryable?: boolean }

export interface AssistantSessionSummary {
  id: number
  user_id: number
  course_id?: number | null
  title: string
  mode: AssistantMode | string
  status: string
  last_message?: string | null
  message_count: number
  created_at: string
  updated_at: string
}

export interface AssistantSessionDetail extends AssistantSessionSummary {
  messages: AssistantMessage[]
}

export interface AssistantSessionListResponse {
  items: AssistantSessionSummary[]
  total: number
  page: number
  page_size: number
}

export interface AssistantSessionCreatePayload {
  course_id?: number | null
  title?: string | null
  mode?: AssistantMode
}

export interface AssistantSendMessagePayload {
  message: string
  course_id?: number | null
  use_course_knowledge?: boolean
  knowledge_document_ids?: number[]
  attachment_file_ids?: number[]
  answer_style?: AssistantAnswerStyle
  top_k?: number
}

export interface AssistantSendMessageResponse {
  session: AssistantSessionSummary
  user_message: AssistantMessage
  assistant_message: AssistantMessage
  answer: string
  references: AssistantReference[]
  used_documents: AssistantUsedDocument[]
  suggested_followups: string[]
  warnings: string[]
}

export interface AssistantFile {
  id: number
  original_filename: string
  content_type?: string | null
  file_size: number
  file_hash: string
  asset_type: string
  parse_status: string
  created_at: string
  updated_at: string
}
